### PR TITLE
update validate designs to report mem type in logs

### PR DESCRIPTION
**Pull request by vishnuchebrolu**
https://jira.xilinx.com/browse/CR-1088690

Validate designs updated to report memory type while reporting throughput.

1. bandwidth_test
**For DDR memory :** 
Throughput (Type: DDR) (Bank count: 4): 52265 MB/s
**For HBM memory :** 
Throughput (Type: HBM) (Bank count: 1): 15328 MB/s
**For DDR and HBM memory :** 
Throughput (Type: DDR) (Bank count: 2): 29265 MB/s
Throughput (Type: HBM) (Bank count: 1): 15328 MB/s 
2. hostmemory_test:
**For HOST memory :** 
Throughput (Type: HOST) (Bank count: 1): 15328 MB/s

@AShivangi Please update the xbutil flow search string to pick the throughput numbers.

### CODE REVIEW

**Review by AShivangi — CHANGES_REQUESTED** [review]
Please make changes as suggested in the email. Thanks!

File: tests/validate/bandwidth_test/src/host.cpp [hunk 1]
```diff
@@ -66,7 +66,8 @@ int main(int argc, char** argv) {
         }
     }
 
-    int NUM_KERNEL;
+    int NUM_KERNEL = 0, NUM_KERNEL_DDR = 0;
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:69]: Please use lower case variable names.  UPPER CASE is reserved for macros (which should not really ever be needed) and would need to be scoped by unique prefix.  Lower case is safe from macro substitution.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 2]
```diff
@@ -77,6 +78,17 @@ int main(int argc, char** argv) {
 
         temp = loadPtreeRoot.get_child("total_ddr_banks");
         NUM_KERNEL = temp.get_value<int>();
+        NUM_KERNEL_DDR = NUM_KERNEL;
+        boost::property_tree::ptree ptMemArray;
+        ptMemArray = loadPtreeRoot.get_child("meminfo");
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:83]: Combine as 
`auto ptMemArray = loadPtreeRoot.get_child("meminfo");`

File: tests/validate/bandwidth_test/src/host.cpp [hunk 3]
```diff
@@ -77,6 +78,17 @@ int main(int argc, char** argv) {
...
+        for (auto memEntry : ptMemArray) {
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:84]: `const auto&`

File: tests/validate/bandwidth_test/src/host.cpp [hunk 4]
```diff
@@ -77,6 +78,17 @@ int main(int argc, char** argv) {
...
+            boost::property_tree::ptree ptMemEntry = memEntry.second;
+            const std::string& sValue = ptMemEntry.get<std::string>("type");
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:86]: `auto sValue = ...;`   You are not getting a reference to any internally held data, so a copy is made any way.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 5]
```diff
@@ -77,6 +78,17 @@ int main(int argc, char** argv) {
...
+            if (sValue == "HBM") {
+                chk_hbm_mem = true;
+            }
+        }
+        if (chk_hbm_mem) NUM_KERNEL_DDR = NUM_KERNEL - 1;
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:91]: Always add a new line (or curly brase if followed by multiple lines) after` if (expression) `

File: tests/validate/bandwidth_test/src/host.cpp [hunk 6]
```diff
@@ -77,6 +78,17 @@ int main(int argc, char** argv) {
...
     } catch (const std::exception& e) {
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:92]: New line before `catch` to line up with` try`

File: tests/validate/bandwidth_test/src/host.cpp [hunk 7]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
 
     double max_throughput = 0;
     int reps = stoi(iter_cnt);
+    if (NUM_KERNEL_DDR) {
+        for (uint32_t i = 4 * 1024; i <= 16 * 1024 * 1024; i *= 2) {
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:165]: This is pure magic (to me that is) :-)

File: tests/validate/bandwidth_test/src/host.cpp [hunk 8]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
+            unsigned int DATA_SIZE = i;
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:166]: lower case variable names

File: tests/validate/bandwidth_test/src/host.cpp [hunk 9]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
 
-    for (uint32_t i = 4 * 1024; i <= 16 * 1024 * 1024; i *= 2) {
-        unsigned int DATA_SIZE = i;
+            if (xcl::is_emulation()) {
+                reps = 2;
+                if (DATA_SIZE > 8 * 1024) break;
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:170]: Magic?
Add  new line after `)`

File: tests/validate/bandwidth_test/src/host.cpp [hunk 10]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
+            unsigned int vector_size_bytes = DATA_SIZE;
+            std::vector<unsigned char, aligned_allocator<unsigned char> > input_host(DATA_SIZE);
+            std::vector<unsigned char, aligned_allocator<unsigned char> > output_host[NUM_KERNEL_DDR];
 
-        unsigned int vector_size_bytes = DATA_SIZE;
-        std::vector<unsigned char, aligned_allocator<unsigned char> > input_host(DATA_SIZE);
-        std::vector<unsigned char, aligned_allocator<unsigned char> > output_host[NUM_KERNEL];
+            for (int i = 0; i < NUM_KERNEL_DDR; i++) {
+                output_host[i].resize(DATA_SIZE);
+            }
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:179]: XRT coding convention is in general
```
if (expression)
  statement;

if (expression) {
  statement1;
  statement2;
}
```

File: tests/validate/bandwidth_test/src/host.cpp [hunk 11]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
 
-        for (int i = 0; i < NUM_KERNEL; i++) {
-            OCL_CHECK(err, err = krnls[i].setArg(0, input_buffer[i]));
-            OCL_CHECK(err, err = krnls[i].setArg(1, output_buffer[i]));
-            OCL_CHECK(err, err = krnls[i].setArg(2, DATA_SIZE));
-            OCL_CHECK(err, err = krnls[i].setArg(3, reps));
+            std::chrono::high_resolution_clock::time_point timeStart;
+            std::chrono::high_resolution_clock::time_point timeEnd;
+
+            timeStart = std::chrono::high_resolution_clock::now();
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:219]: `auto time_start = ...;`

File: tests/validate/bandwidth_test/src/host.cpp [hunk 12]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
+                OCL_CHECK(err, err = q.finish());
+            }
 
-        for (int i = 0; i < NUM_KERNEL; i++) {
-            OCL_CHECK(err, err = krnls[i].setArg(0, input_buffer[i]));
-            OCL_CHECK(err, err = krnls[i].setArg(1, output_buffer[i]));
-            OCL_CHECK(err, err = krnls[i].setArg(2, DATA_SIZE));
-            OCL_CHECK(err, err = krnls[i].setArg(3, reps));
+            std::chrono::high_resolution_clock::time_point timeStart;
+            std::chrono::high_resolution_clock::time_point timeEnd;
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:217]: remove

File: tests/validate/bandwidth_test/src/host.cpp [hunk 13]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
+
+            timeStart = std::chrono::high_resolution_clock::now();
+
+            for (int i = 0; i < NUM_KERNEL_DDR; i++) {
+                OCL_CHECK(err, err = q.enqueueTask(krnls[i]));
+            }
+            q.finish();
+            timeEnd = std::chrono::high_resolution_clock::now();
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:225]: auto time_end = ...;

File: tests/validate/bandwidth_test/src/host.cpp [hunk 14]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
+                        return EXIT_FAILURE;
+                    }
+                }
+            }
+
+            double usduration;
+            double dnsduration;
+            double dsduration;
+            double bpersec;
+            double mbpersec;
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:248]: Combine with usage

File: tests/validate/bandwidth_test/src/host.cpp [hunk 15]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
+
+            if (mbpersec > max_throughput) max_throughput = mbpersec;
         }
 
-        for (int i = 0; i < NUM_KERNEL; i++) {
-            OCL_CHECK(err, err = q.enqueueWriteBuffer(input_buffer[i], CL_TRUE, 0, vector_size_bytes, input_host.data(),
+        std::cout << "Overall DDRs (total " << NUM_KERNEL_DDR << ") Throughput: " << max_throughput << "MB/s\n";
+    }
+    if (chk_hbm_mem) {
+        for (uint32_t i = 4 * 1024; i <= 16 * 1024 * 1024; i *= 2) {
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:264]: magic?

File: tests/validate/bandwidth_test/src/host.cpp [hunk 16]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
             OCL_CHECK(err, err = q.finish());
-        }
 
-        std::chrono::high_resolution_clock::time_point timeStart;
-        std::chrono::high_resolution_clock::time_point timeEnd;
+            std::chrono::high_resolution_clock::time_point timeStart;
+            std::chrono::high_resolution_clock::time_point timeEnd;
 
-        timeStart = std::chrono::high_resolution_clock::now();
+            timeStart = std::chrono::high_resolution_clock::now();
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:304]: As earlier.  and don't mix and match camelCase with under_bar_names.   My preference is latter, but consistency is what is important.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 17]
```diff
@@ -151,101 +161,184 @@ int main(int argc, char** argv) {
...
-        double usduration;
-        double dnsduration;
-        double dsduration;
-        double bpersec;
-        double mbpersec;
+            double usduration;
+            double dnsduration;
+            double dsduration;
+            double bpersec;
+            double mbpersec;
```
Comment (stsoe) [inline tests/validate/bandwidth_test/src/host.cpp:327]: Combine with usage.    It all looks like deja vu, I didn't check, but is refactoring possible?

**Review by stsoe — CHANGES_REQUESTED** [review]

File: tests/validate/bandwidth_test/src/host.cpp (hunk 15 above)
Comment (vishnuchebrolu) [reply tests/validate/bandwidth_test/src/host.cpp:264]: We are trying to run with different buffer sizes(4KB -> 16MB).

**Comment by vishnuchebrolu** [comment]
> Please make changes as suggested in the email. Thanks!

@AShivangi  I have made the suggested changes as per email.
Can you please merge this PR along with following https://github.com/Xilinx/XRT/pull/5793 to overcome conflicts with naming from slavebridge -> hostmemory.

**Review by stsoe — APPROVED** [review]
I trust changes were made.

**Comment by AShivangi** [comment]
@vishnuchebrolu, Please update the PR desc to reflect your latest changes for documentation purposes. I'll approve the PR once you do so.

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 18]
```diff
@@ -349,11 +349,21 @@ runTestCase( const std::shared_ptr<xrt_core::device>& _dev, const std::string& p
...
   if(xclbin.compare("bandwidth.xclbin") == 0) {
+    // old testcases where we have "Maximum throughput:"
     size_t st = os_stdout.str().find("Maximum");
     if (st != std::string::npos) {
       size_t end = os_stdout.str().find("\n", st);
       logger(_ptTest, "Details", os_stdout.str().substr(st, end - st));
     }
+    else {
+      // new test cases to find "Throughput (Type: {...}) (Bank count: {...}):"
+      auto st = os_stdout.str().find("Throughput", 0);
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:360]: Post 2021.2, we should find a better way to communicate this information back to xbutil.   "Screen Scraping" for the "Throughput" keyword of the console output is not the way it should be done.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 19]
```diff
@@ -54,9 +54,9 @@ int main(int argc, char** argv) {
         std::cout << "ERROR : please provide the platform test path to -p option\n";
         return EXIT_FAILURE;
     }
+    std::string binaryFile = test_path + b_file;
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:57]: Please use boost::filesystem::path to join a path and file.  The solution provided above will be problematic across multiple OSs.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 20]
```diff
@@ -66,26 +66,36 @@ int main(int argc, char** argv) {
...
+        }
+        if (chk_hbm_mem) {
+            num_kernel_ddr = num_kernel - 1;
+        }
     } catch (const std::exception& e) {
         std::string msg("ERROR: Bad JSON format detected while marshaling build metadata (");
         msg += e.what();
         msg += ").";
         std::cout << msg << std::endl;
     }
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:97]: Hmm...If an error was found, shouldn't the code stop testing this test by throwing an exception?

File: tests/validate/bandwidth_test/src/host.cpp [hunk 21]
```diff
@@ -151,102 +159,171 @@ int main(int argc, char** argv) {
 
     double max_throughput = 0;
     int reps = stoi(iter_cnt);
+    if (num_kernel_ddr) {
+        for (uint32_t i = 4 * 1024; i <= 16 * 1024 * 1024; i *= 2) {
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:163]: Please add a comment on what you are doing here.  For example:
Starting at 4K and going up to 16M with increments of power of 2 ....

File: tests/validate/bandwidth_test/src/host.cpp [hunk 22]
```diff
@@ -151,102 +159,171 @@ int main(int argc, char** argv) {
...
+            unsigned int data_size = i;
+
+            if (xcl::is_emulation()) {
+                reps = 2;
+                if (data_size > 8 * 1024) {
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:168]: What is so magical about 8K, please add a comment.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 23]
```diff
@@ -151,102 +159,171 @@ int main(int argc, char** argv) {
...
+                OCL_CHECK(err,
+                          output_buffer[i] = cl::Buffer(context, CL_MEM_READ_WRITE, vector_size_bytes, nullptr, &err));
+            }
 
-        // Initializing output vectors to zero
-        for (int i = 0; i < NUM_KERNEL; i++) {
-            std::fill(output_host[i].begin(), output_host[i].end(), 0);
-        }
+            for (int i = 0; i < num_kernel_ddr; i++) {
+                OCL_CHECK(err, err = krnls[i].setArg(0, input_buffer[i]));
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:204]: Document the arguments on what is being set.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 24]
```diff
@@ -151,102 +159,171 @@ int main(int argc, char** argv) {
...
+                        return EXIT_FAILURE;
+                    }
+                }
+            }
+
+            double usduration =
+                (double)(std::chrono::duration_cast<std::chrono::nanoseconds>(time_end - time_start).count() / reps);
+
+            double dnsduration = (double)usduration;
+            double dsduration = dnsduration / ((double)1000000000);
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:245]: Magic number....what is the expected duration, seconds, msec, etc.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 25]
```diff
@@ -151,102 +159,171 @@ int main(int argc, char** argv) {
...
+            double bpersec = (data_size * num_kernel_ddr) / dsduration;
+            double mbpersec = (2 * bpersec) / ((double)1024 * 1024); // For concurrent Read/Write
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:247]: Lots of magic values.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 26]
```diff
@@ -151,102 +159,171 @@ int main(int argc, char** argv) {
...
+
+            if (mbpersec > max_throughput) max_throughput = mbpersec;
         }
 
-        for (int i = 0; i < NUM_KERNEL; i++) {
-            OCL_CHECK(err, err = q.enqueueWriteBuffer(input_buffer[i], CL_TRUE, 0, vector_size_bytes, input_host.data(),
+        std::cout << "Throughput (Type: DDR) (Bank count: " << num_kernel_ddr << ") : " << max_throughput << "MB/s\n";
+    }
+    if (chk_hbm_mem) {
+        for (uint32_t i = 4 * 1024; i <= 16 * 1024 * 1024; i *= 2) {
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:255]: Please document what this for loop is doing (see earlier comment).

File: tests/validate/bandwidth_test/src/host.cpp [hunk 27]
```diff
@@ -151,102 +159,171 @@ int main(int argc, char** argv) {
...
+            unsigned int data_size = i;
+
+            if (xcl::is_emulation()) {
+                reps = 2;
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:259]: Magic numbers

File: tests/validate/bandwidth_test/src/host.cpp [hunk 28]
```diff
@@ -151,102 +159,171 @@ int main(int argc, char** argv) {
...
-        double bpersec;
-        double mbpersec;
+            double usduration =
+                (double)(std::chrono::duration_cast<std::chrono::nanoseconds>(time_end - time_start).count() / reps);
 
-        usduration = (double)(std::chrono::duration_cast<std::chrono::nanoseconds>(timeEnd - timeStart).count() / reps);
+            double dnsduration = (double)usduration;
+            double dsduration = dnsduration / ((double)1000000000);
+            double bpersec = data_size / dsduration;
+            double mbpersec = (2 * bpersec) / ((double)1024 * 1024); // For concurrent Read/Write
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:319]: Magic numbers.

File: tests/validate/bandwidth_test/src/host.cpp [hunk 29]
```diff
@@ -54,9 +54,9 @@ int main(int argc, char** argv) {
         std::cout << "ERROR : please provide the platform test path to -p option\n";
```
Comment (rozumx) [inline tests/validate/bandwidth_test/src/host.cpp:54]: This main() function is huge with multiple duplicate sections,  please consider refactoring it into smaller functions.

File: tests/validate/slavebridge_test/src/host.cpp [hunk 30]
```diff
@@ -53,58 +54,73 @@ int main(int argc, char** argv) {
         std::cout << "ERROR : please provide the platform test path to -p option\n";
         return EXIT_FAILURE;
     }
+    std::string binary_file = test_path + b_file;
```
Comment (rozumx) [inline tests/validate/slavebridge_test/src/host.cpp:57]: Please use boost::filesystem::path to join paths and files.

File: tests/validate/slavebridge_test/src/host.cpp [hunk 31]
```diff
@@ -53,58 +54,73 @@ int main(int argc, char** argv) {
...
         } else {
             std::cout << "\nSUPPORTED" << std::endl;
             return EXIT_SUCCESS;
         }
     }
 
-    int NUM_KERNEL;
+    int num_kernel;
     std::string filename = "/platform.json";
     std::string platform_json = test_path + filename;
```
Comment (rozumx) [inline tests/validate/slavebridge_test/src/host.cpp:79]: please use boost::filesystem::paths to join paths and file names.

**Review by rozumx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp (hunk 18 above)
Comment (AShivangi) [reply src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:360]: I agree

**Comment by vishnuchebrolu** [comment]
> @vishnuchebrolu, Please update the PR desc to reflect your latest changes for documentation purposes. I'll approve the PR once you do so.

@AShivangi I have updated the PR description. Also taken care of the comments from stephen.

**Comment by vishnuchebrolu** [comment]
@AShivangi  As there are conflicts in this PR I have made all the requested changes by Soren and Stephen in new PR.
https://github.com/Xilinx/XRT/pull/5829

**Pull request closed without merging**